File: SeLeAs/Python/SmallestInfiniteSet.py
import heapq

class SmallestInfiniteSet:

    def __init__(self):
        self.current = 1
        self.added_back = []  # min-heap
        self.added_back_set = set()  # to avoid duplicates

    def popSmallest(self) -> int:
        if self.added_back:
            smallest = heapq.heappop(self.added_back)
            self.added_back_set.remove(smallest)
            return smallest
        else:
            self.current += 1
            return self.current - 1

    def addBack(self, num: int) -> None:
        if num < self.current and num not in self.added_back_set:
            heapq.heappush(self.added_back, num)
            self.added_back_set.add(num)

File: SeLeAs/LeetCode/LeetCode_75/SmallestInfiniteSet.java
import java.util.PriorityQueue;
import java.util.HashSet;
public class SmallestInfiniteSet {




    private int current;
    private PriorityQueue<Integer> addedBack;
    private HashSet<Integer> addedBackSet;

    public SmallestInfiniteSet() {
        current = 1;
        addedBack = new PriorityQueue<>();
        addedBackSet = new HashSet<>();
    }

    public int popSmallest() {
        if (!addedBack.isEmpty()) {
            int smallest = addedBack.poll();
            addedBackSet.remove(smallest);
            return smallest;
        } else {
            return current++;
        }
    }

    public void addBack(int num) {
        if (num < current && !addedBackSet.contains(num)) {
            addedBack.add(num);
            addedBackSet.add(num);
        }
    }
}

